zenoh-ext/src/advanced_subscriber.rs

```diff
@@ -0,0 +1,1316 @@
+            let query_target = states.query_target;
+            let query_timeout = states.query_timeout;
+            drop(lock);
+            let handler = SequencedRepliesHandler {
+                source_id: self.source_id,
+                statesref: self.statesref.clone(),
+            };
+            let _ = session
+                .get(Selector::from((query_expr, seq_num_range)))
+                .callback({
+                    move |r: Reply| {
+                        if let Ok(s) = r.into_result() {
+                            if key_expr.intersects(s.key_expr()) {
+                                let states = &mut *zlock!(handler.statesref);
+                                handle_sample(states, s);
+                            }
+                        }
+                    }
+                })
+                .consolidation(ConsolidationMode::None)
+                .accept_replies(ReplyKeyExpr::Any)
+                .target(query_target)
+                .timeout(query_timeout)
+                .wait();
+        }
+    }
+}
+
+#[zenoh_macros::unstable]
+impl<Handler> AdvancedSubscriber<Handler> {
+    fn new<H>(conf: AdvancedSubscriberBuilder<'_, '_, '_, H>) -> ZResult<Self>
+    where
+        H: IntoHandler<Sample, Handler = Handler> + Send,
+    {
+        let (callback, receiver) = conf.handler.into_handler();
+        let key_expr = conf.key_expr?;
+        let meta = match conf.meta_key_expr {
+            Some(meta) => Some(meta?),
+            None => None,
+        };
+        let retransmission = conf.retransmission;
+        let query_target = conf.query_target;
+        let query_timeout = conf.query_timeout;
+        let session = conf.session.clone();
+        let statesref = Arc::new(Mutex::new(State {
+            next_id: 0,
+            sequenced_states: HashMap::new(),
+            timestamped_states: HashMap::new(),
+            global_pending_queries: if conf.history.is_some() { 1 } else { 0 },
+            session,
+            period: retransmission.as_ref().and_then(|r| {
+                r.periodic_queries.map(|p| Period {
+                    timer: Timer::new(false),
+                    period: p,
+                })
+            }),
+            key_expr: key_expr.clone().into_owned(),
+            retransmission: retransmission.is_some(),
+            query_target: conf.query_target,
+            query_timeout: conf.query_timeout,
+            callback: callback.clone(),
+            miss_handlers: HashMap::new(),
+        }));
+
+        let sub_callback = {
+            let statesref = statesref.clone();
+            let session = conf.session.clone();
+            let key_expr = key_expr.clone().into_owned();
+
+            move |s: Sample| {
+                let mut lock = zlock!(statesref);
+                let states = &mut *lock;
+                let source_id = s.source_info().source_id().cloned();
+                let new = handle_sample(states, s);
```

Comment:
Samples are handled with lock held. If there are multiple connections, i.e. P2P, it samples from different publishers to be handled in parallel, and it will block the transports.
Having only one big lock, while publishers are independently managed is a big issue.